internal/provider/waypoint/resource_waypoint_application.go

```diff
@@ -0,0 +1,426 @@
+var _ resource.Resource = &ApplicationResource{}
+var _ resource.ResourceWithImportState = &ApplicationResource{}
+
+func NewApplicationResource() resource.Resource {
+	return &ApplicationResource{}
+}
+
+// ApplicationResource defines the resource implementation.
+type ApplicationResource struct {
+	client *clients.Client
+}
+
+// ApplicationResourceModel describes the resource data model.
+type ApplicationResourceModel struct {
+	ID                      types.String `tfsdk:"id"`
+	Name                    types.String `tfsdk:"name"`
+	ProjectID               types.String `tfsdk:"project_id"`
+	OrgID                   types.String `tfsdk:"organization_id"`
+	ReadmeMarkdown          types.String `tfsdk:"readme_markdown"`
+	ApplicationTemplateID   types.String `tfsdk:"application_template_id"`
+	ApplicationTemplateName types.String `tfsdk:"application_template_name"`
+	NamespaceID             types.String `tfsdk:"namespace_id"`
+
+	// deferred for now
+	// Tags       types.List `tfsdk:"tags"`
+
+	// deferred and probably a list or objects, but may possible be a separate
+	// ActionCfgs types.List `tfsdk:"action_cfgs"`
+}
+
+func (r *ApplicationResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_waypoint_application"
+}
+
+func (r *ApplicationResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "Waypoint Application resource",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "The ID of the Application.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the Application.",
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"organization_id": schema.StringAttribute{
+				Description: "The ID of the HCP organization where the Waypoint Application is located.",
+				Computed:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"project_id": schema.StringAttribute{
+				Description: "The ID of the HCP project where the Waypoint Application is located.",
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"application_template_id": schema.StringAttribute{
+				Required:    true,
+				Description: "ID of the Application Template this Application is based on.",
+			},
+			// application_template_name is a computed only attribute for ease
+			// of reference
+			"application_template_name": schema.StringAttribute{
```

Comment:
May need to change this name to `template_name`, pending the outcome of our other comment thread on this PR w/@briancain.

Reply:
If we do, we'll do it in another PR that updates it across the board.